contracts/utils/MultiRateLimited.sol

```diff
@@ -0,0 +1,368 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import {CoreRef} from "../refs/CoreRef.sol";
+import {TribeRoles} from "./../core/TribeRoles.sol";
+import {RateLimited} from "./RateLimited.sol";
+import {IMultiRateLimited} from "./IMultiRateLimited.sol";
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+/// @title abstract contract for putting a rate limit on how fast an address can perform an action e.g. Minting
+/// there are two buffers, one buffer which is each individual addresses's current buffer,
+/// and then there is a global buffer which is the buffer that each individual address must respect as well
+/// @author Elliot Friedman, Fei Protocol
+/// this contract was made abstract so that other contracts that already construct an instance of CoreRef
+/// do not collide with this one
+abstract contract MultiRateLimited is RateLimited, IMultiRateLimited {
+    using SafeCast for *;
+
+    /// @notice the struct containing all information per rate limited address
+    struct RateLimitData {
+        uint32 lastBufferUsedTime;
+        uint112 bufferCap;
+        uint112 bufferStored;
+        uint112 rateLimitPerSecond;
+    }
+
+    /// @notice rate limited address information
+    mapping(address => RateLimitData) public rateLimitPerAddress;
+
+    /// @notice max rate limit per second allowable by non governor per contract
+    uint256 public individualMaxRateLimitPerSecond;
+
+    /// @notice max buffer cap allowable by non governor per contract
+    uint256 public individualMaxBufferCap;
+
+    /// @param _maxRateLimitPerSecond maximum amount of fei that can replenish per second ever, this amount cannot be changed by governance
+    /// @param _rateLimitPerSecond maximum rate limit per second per address
+    /// @param _individualMaxRateLimitPerSecond maximum rate limit per second per address in multi rate limited
+    /// @param _individualMaxBufferCap maximum buffer cap in multi rate limited
+    /// @param _globalBufferCap maximum global buffer cap
+    constructor(
+        uint256 _maxRateLimitPerSecond,
+        uint256 _rateLimitPerSecond,
+        uint256 _individualMaxRateLimitPerSecond,
+        uint256 _individualMaxBufferCap,
+        uint256 _globalBufferCap
+    )
+        RateLimited(
+            _maxRateLimitPerSecond,
+            _rateLimitPerSecond,
+            _globalBufferCap,
+            false
```

Comment:
you may want to pull in RateLimitedV2 here as it could save lots of gas in the system

Reply:
It also wouldn't require you to set the `doPartialAction` bool to false

Reply:
RateLimitedV2 isn't audited :(